go/types/opcache.go

```diff
@@ -91,37 +125,100 @@ func encToSlice(v Value, initBuf []byte, vw ValueWriter) []byte {
 	return w.data()
 }
 
-func (p *opCache) NewIterator() *opCacheIterator {
-	return &opCacheIterator{p.ops.NewIterator(nil, nil), p.vrw}
+func (opc *ldbOpCache) NewIterator() opCacheIterator {
+	prefix := [4]byte{}
+	binary.LittleEndian.PutUint32(prefix[:], opc.colId)
+	return &ldbOpCacheIterator{iter: opc.store.ldb.NewIterator(util.BytesPrefix(prefix[:]), nil), vr: opc.vrw}
 }
 
-func (p *opCache) Destroy() error {
-	d.Chk.NoError(p.ops.Close())
-	return os.RemoveAll(p.dbDir)
-}
-
-func (i *opCacheIterator) Next() bool {
+func (i *ldbOpCacheIterator) Next() bool {
 	return i.iter.Next()
 }
 
-func (i *opCacheIterator) Op() sequenceItem {
+func (i *ldbOpCacheIterator) MapOp() sequenceItem {
 	entry := mapEntry{}
 	ldbKey := i.iter.Key()
-	data := i.iter.Value()
-	dataOffset := 0
-	switch NomsKind(ldbKey[0]) {
+	ldbValue := i.iter.Value()
+	switch NomsKind(ldbKey[uint32Size]) {
 	case BoolKind, NumberKind, StringKind:
-		entry.key = DecodeFromBytes(ldbKey, i.vr, staticTypeCache)
+		entry.key = DecodeFromBytes(ldbKey[4:], i.vr, staticTypeCache)
+		entry.value = DecodeFromBytes(ldbValue, i.vr, staticTypeCache)
 	default:
-		keyBytesLen := int(binary.LittleEndian.Uint32(data))
-		entry.key = DecodeFromBytes(data[uint32Size:uint32Size+keyBytesLen], i.vr, staticTypeCache)
-		dataOffset = uint32Size + keyBytesLen
+		keyBytesLen := int(binary.LittleEndian.Uint32(ldbValue))
+		entry.key = DecodeFromBytes(ldbValue[uint32Size:uint32Size+keyBytesLen], i.vr, staticTypeCache)
+		entry.value = DecodeFromBytes(ldbValue[uint32Size+keyBytesLen:], i.vr, staticTypeCache)
 	}
 
-	entry.value = DecodeFromBytes(data[dataOffset:], i.vr, staticTypeCache)
 	return entry
 }
 
-func (i *opCacheIterator) Release() {
+// Insert can be called from any goroutine
+func (opc *ldbOpCache) SetInsert(val Value) {
+
+	switch val.Type().Kind() {
+	default:
+		ldbKey := ldbKeyFromValueHash(val, opc.colId)
+		valArray := [initialBufferSize]byte{}
+		encodedVal := encToSlice(val, valArray[:], opc.vrw)
+		err := opc.store.ldb.Put(ldbKey, encodedVal, nil)
+		d.Chk.NoError(err)
+
+	case BoolKind, NumberKind, StringKind:
+		ldbKey := ldbKeyFromValue(val, opc.colId, opc.vrw)
+		// Since the ldbKey contains the val, there's no reason to store anything in the ldbValue
+		err := opc.store.ldb.Put(ldbKey, nil, nil)
+		d.Chk.NoError(err)
+	}
+}
+
+func (i *ldbOpCacheIterator) SetOp() sequenceItem {
+	ldbKey := i.iter.Key()
+
+	switch NomsKind(ldbKey[uint32Size]) {
+	case BoolKind, NumberKind, StringKind:
+		return DecodeFromBytes(ldbKey[uint32Size:], i.vr, staticTypeCache)
```

Comment:
fyi, this particular issue caused me a few hours in debugging time. There's a comment about this on ldbKeyFromValue(). This is the way things were encoded originally and I opted to maintain compatibility with opCacheComparer rather than change the encoding by adding an additional byte here.